File: Assets/Scripts/Battle/AstralBattleEngine.cs
using System;
using System.Collections.Generic;

namespace AstralWilds
{
    /// <summary>
    /// Owns the demo's 2v2 battle-round mechanics: queuing player actions (attack,
    /// Arc Burst, Guard), applying damage, and resolving each opponent
    /// counterattack round. Extracted from AstralDemoLoopController so the
    /// mechanics are unit-testable without a live Unity scene. Flow-state
    /// transitions (Encounter -> Battle -> Recruitment/Defeat), messages, save
    /// data, and rewards remain the controller's responsibility; this class only
    /// answers "what happened" so the controller can narrate and transition.
    /// </summary>
    public sealed class AstralBattleEngine
    {
        public enum ActionOutcome { Invalid, SlotAlreadyActed, Applied }

        public readonly struct RoundOutcome
        {
            public RoundOutcome(bool resolved, bool playerDefeat, string summary)
            {
                Resolved = resolved;
                PlayerDefeat = playerDefeat;
                Summary = summary;
            }

            /// <summary>False when one active slot still needs to act this round.</summary>
            public bool Resolved { get; }
            public bool PlayerDefeat { get; }
            public string Summary { get; }
        }

        private readonly IReadOnlyList<AstralDemoMember> party;
        private readonly int[] activeParty;
        private readonly AstralDemoMember[] opponents = new AstralDemoMember[2];
        private readonly bool[] acted = new bool[2];
        private readonly bool[] guarded = new bool[2];
        private readonly int opponentDamage;
        public AstralBattleState Battle { get; }

        public AstralBattleEngine(
            IReadOnlyList<AstralDemoMember> party,
            int[] activeParty,
            string primaryId, string primaryName,
            string companionId, string companionName,
            int opponentDamage)
        {
            this.party = party ?? throw new ArgumentNullException(nameof(party));
            this.activeParty = activeParty ?? throw new ArgumentNullException(nameof(activeParty));
            if (activeParty.Length != 2)
                throw new ArgumentException("Exactly two active slots are required.", nameof(activeParty));
            this.opponentDamage = Math.Max(1, opponentDamage);

            opponents[0] = new AstralDemoMember { id = primaryId, displayName = primaryName };
            opponents[1] = new AstralDemoMember { id = companionId, displayName = companionName };

            var playerParty = new AstralParty(AstralParty.PlayerCapacity);
            foreach (var member in party)
            {
                var combatant = new AstralCombatant(member.id, member.displayName);
                if (member.defeated) combatant.MarkDefeated();
                playerParty.TryAdd(combatant);
            }
            var enemyParty = new AstralParty(AstralParty.OpponentCapacity);
            foreach (var enemy in opponents)
                enemyParty.TryAdd(new AstralCombatant(enemy.id, enemy.displayName));

            Battle = new AstralBattleState(playerParty, enemyParty);
            for (int slot = 0; slot < 2; slot++)
            {
                Battle.TryActivate(BattleSide.Player, activeParty[slot], slot);
                Battle.TryActivate(BattleSide.Opponent, slot, slot);
            }
        }

        public AstralDemoMember GetOpponent(int slot) => (slot >= 0 && slot < 2) ? opponents[slot] : null;
        public bool HasActed(int slot) => slot >= 0 && slot < 2 && acted[slot];
        public bool IsGuarded(int slot) => slot >= 0 && slot < 2 && guarded[slot];

        public AstralDemoMember GetActiveParty(int slot)
        {
            if (slot < 0 || slot >= activeParty.Length || activeParty[slot] < 0 || activeParty[slot] >= party.Count)
                return null;
            return party[activeParty[slot]];
        }

        public bool AllOpponentsDefeated() => opponents[0] == null || (opponents[0].defeated && opponents[1].defeated);

        public ActionOutcome QueueAttack(int actorSlot, int targetSlot, out int damageDealt)
        {
            damageDealt = 0;
            AstralDemoMember actor = GetActiveParty(actorSlot);
            AstralDemoMember target = GetOpponent(targetSlot);
            if (actor == null || actor.defeated || target == null || target.defeated)
                return ActionOutcome.Invalid;
            if (acted[actorSlot] || !Battle.TryQueueAction(BattleSide.Player, actorSlot,
                new QueuedAstralAction("attack", TargetScope.OneEnemy, targetSlot)))
                return ActionOutcome.SlotAlreadyActed;

            acted[actorSlot] = true;
            damageDealt = AstralCombatRules.BasicAttackDamage;
            ApplyDamageToOpponent(targetSlot, damageDealt);
            return ActionOutcome.Applied;
        }

        public ActionOutcome QueueArcBurst(int actorSlot, out int targetsHit)
        {
            targetsHit = 0;
            AstralDemoMember actor = GetActiveParty(actorSlot);
            if (actor == null || actor.defeated)
                return ActionOutcome.Invalid;
            if (acted[actorSlot] || !Battle.TryQueueAction(BattleSide.Player, actorSlot,
                new QueuedAstralAction("arc-burst", TargetScope.BothEnemies)))
                return ActionOutcome.SlotAlreadyActed;

            acted[actorSlot] = true;
            for (int slot = 0; slot < opponents.Length; slot++)
            {
                if (opponents[slot] == null || opponents[slot].defeated)
                    continue;
                ApplyDamageToOpponent(slot, AstralCombatRules.ArcBurstDamagePerTarget);
                targetsHit++;
            }
            return ActionOutcome.Applied;
        }

        public ActionOutcome QueueGuard(int actorSlot)
        {
            AstralDemoMember actor = GetActiveParty(actorSlot);
            if (actor == null || actor.defeated)
                return ActionOutcome.Invalid;
            if (acted[actorSlot] || !Battle.TryQueueAction(BattleSide.Player, actorSlot,
                new QueuedAstralAction("guard", TargetScope.Self, actorSlot)))
                return ActionOutcome.SlotAlreadyActed;

            acted[actorSlot] = true;
            guarded[actorSlot] = true;
            return ActionOutcome.Applied;
        }

        /// <summary>
        /// Marks the actor slot as acted without a battle action (used by a
        /// voluntary bench swap, which still consumes the slot's turn).
        /// </summary>
        public void MarkActed(int slot)
        {
            if (slot >= 0 && slot < 2)
                acted[slot] = true;
        }

        private void ApplyDamageToOpponent(int slot, int damage)
        {
            AstralDemoMember target = opponents[slot];
            if (target == null || target.defeated || damage <= 0)
                return;

            target.hp = Math.Max(0, target.hp - damage);
            if (target.hp == 0)
            {
                target.defeated = true;
                Battle.OpponentParty.Astrals[slot].MarkDefeated();
            }
        }

        /// <summary>
        /// Resolves the round once both active party slots have acted (or are
        /// unable to). Returns Resolved=false if a slot is still waiting.
        /// </summary>
        public RoundOutcome TryFinishRound()
        {
            for (int slot = 0; slot < 2; slot++)
            {
                AstralDemoMember member = GetActiveParty(slot);
                if (member != null && !member.defeated && !acted[slot])
                    return new RoundOutcome(false, false, null);
            }

            string summary = ResolveOpponentActions();
            acted[0] = acted[1] = false;
            Battle.ClearQueuedActions();

            bool playerDefeat = AllDefeated(party);
            return new RoundOutcome(true, playerDefeat, summary);
        }

        private string ResolveOpponentActions()
        {
            var results = new List<string>(2);
            for (int i = 0; i < opponents.Length; i++)
            {
                AstralDemoMember enemy = opponents[i];
                int targetSlot = i % 2;
                AstralDemoMember target = GetActiveParty(targetSlot);
                if (target == null || target.defeated)
                {
                    targetSlot = (i + 1) % 2;
                    target = GetActiveParty(targetSlot);
                }
                if (enemy == null || enemy.defeated || target == null || target.defeated)
                    continue;

                bool wasGuarded = guarded[targetSlot];
                int damage = AstralCombatRules.ResolveIncomingDamage(opponentDamage, wasGuarded);
                target.hp = Math.Max(0, target.hp - damage);
                if (target.hp == 0)
                {
                    target.defeated = true;
                    int partyIndex = IndexOfMember(target);
                    if (partyIndex >= 0)
                        Battle.PlayerParty.Astrals[partyIndex].MarkDefeated();
                }
                results.Add($"{target.displayName}{(wasGuarded ? " guarded and" : "")} took {damage}{(target.defeated ? " and fainted" : "")}");
            }
            guarded[0] = guarded[1] = false;
            return results.Count == 0 ? "no counterattacks landed." : string.Join("; ", results) + ".";
        }

        private int IndexOfMember(AstralDemoMember member)
        {
            for (int i = 0; i < party.Count; i++)
                if (ReferenceEquals(party[i], member))
                    return i;
            return -1;
        }

        public static int FindFirstEligibleParty(IReadOnlyList<AstralDemoMember> party, int start)
        {
            for (int i = Math.Max(0, start); i < party.Count; i++)
                if (!party[i].defeated)
                    return i;
            return -1;
        }

        public static bool AllDefeated(IReadOnlyList<AstralDemoMember> party)
        {
            for (int i = 0; i < party.Count; i++)
                if (!party[i].defeated)
                    return false;
            return true;
        }
    }
}

File: Astral_Wilds_Unreal/Astral_Wilds/Source/Astral_Wilds/AstralBattleEngine.cpp
#include "AstralBattleEngine.h"
#include "AstralCombatRules.h"

void UAstralBattleEngine::Initialize(TArray<FAstralCombatant>* InParty, const int32 InActiveParty[2],
	const FString& PrimaryId, const FString& PrimaryName,
	const FString& CompanionId, const FString& CompanionName,
	int32 InOpponentDamage)
{
	Party = InParty;
	ActiveParty[0] = InActiveParty[0];
	ActiveParty[1] = InActiveParty[1];
	OpponentDamage = FMath::Max(1, InOpponentDamage);

	Opponents[0] = FAstralCombatant();
	Opponents[0].Id = PrimaryId;
	Opponents[0].DisplayName = PrimaryName;

	Opponents[1] = FAstralCombatant();
	Opponents[1].Id = CompanionId;
	Opponents[1].DisplayName = CompanionName;

	bActed[0] = bActed[1] = false;
	bGuarded[0] = bGuarded[1] = false;
}

FAstralCombatant UAstralBattleEngine::GetOpponent(int32 Slot) const
{
	if (Slot < 0 || Slot > 1)
	{
		return FAstralCombatant();
	}
	return Opponents[Slot];
}

bool UAstralBattleEngine::HasActed(int32 Slot) const
{
	return Slot >= 0 && Slot < 2 && bActed[Slot];
}

bool UAstralBattleEngine::IsGuarded(int32 Slot) const
{
	return Slot >= 0 && Slot < 2 && bGuarded[Slot];
}

FAstralCombatant* UAstralBattleEngine::GetActiveParty(int32 Slot) const
{
	if (!Party || Slot < 0 || Slot > 1)
	{
		return nullptr;
	}
	const int32 PartyIndex = ActiveParty[Slot];
	if (PartyIndex < 0 || PartyIndex >= Party->Num())
	{
		return nullptr;
	}
	return &(*Party)[PartyIndex];
}

bool UAstralBattleEngine::AllOpponentsDefeated() const
{
	return Opponents[0].bDefeated && Opponents[1].bDefeated;
}

EAstralActionOutcome UAstralBattleEngine::QueueAttack(int32 ActorSlot, int32 TargetSlot, int32& OutDamageDealt)
{
	OutDamageDealt = 0;
	FAstralCombatant* Actor = GetActiveParty(ActorSlot);
	if (TargetSlot < 0 || TargetSlot > 1)
	{
		return EAstralActionOutcome::Invalid;
	}
	FAstralCombatant& Target = Opponents[TargetSlot];

	if (!Actor || Actor->bDefeated || Target.bDefeated)
	{
		return EAstralActionOutcome::Invalid;
	}
	if (bActed[ActorSlot])
	{
		return EAstralActionOutcome::SlotAlreadyActed;
	}

	bActed[ActorSlot] = true;
	OutDamageDealt = UAstralCombatRules::BasicAttackDamage;
	ApplyDamageToOpponent(TargetSlot, OutDamageDealt);
	return EAstralActionOutcome::Applied;
}

EAstralActionOutcome UAstralBattleEngine::QueueArcBurst(int32 ActorSlot, int32& OutTargetsHit)
{
	OutTargetsHit = 0;
	FAstralCombatant* Actor = GetActiveParty(ActorSlot);
	if (!Actor || Actor->bDefeated)
	{
		return EAstralActionOutcome::Invalid;
	}
	if (bActed[ActorSlot])
	{
		return EAstralActionOutcome::SlotAlreadyActed;
	}

	bActed[ActorSlot] = true;
	for (int32 Slot = 0; Slot < 2; Slot++)
	{
		if (Opponents[Slot].bDefeated)
		{
			continue;
		}
		ApplyDamageToOpponent(Slot, UAstralCombatRules::ArcBurstDamagePerTarget);
		OutTargetsHit++;
	}
	return EAstralActionOutcome::Applied;
}

EAstralActionOutcome UAstralBattleEngine::QueueGuard(int32 ActorSlot)
{
	FAstralCombatant* Actor = GetActiveParty(ActorSlot);
	if (!Actor || Actor->bDefeated)
	{
		return EAstralActionOutcome::Invalid;
	}
	if (bActed[ActorSlot])
	{
		return EAstralActionOutcome::SlotAlreadyActed;
	}

	bActed[ActorSlot] = true;
	bGuarded[ActorSlot] = true;
	return EAstralActionOutcome::Applied;
}

void UAstralBattleEngine::MarkActed(int32 Slot)
{
	if (Slot >= 0 && Slot < 2)
	{
		bActed[Slot] = true;
	}
}

void UAstralBattleEngine::ApplyDamageToOpponent(int32 Slot, int32 Damage)
{
	if (Slot < 0 || Slot > 1)
	{
		return;
	}
	FAstralCombatant& Target = Opponents[Slot];
	if (Target.bDefeated || Damage <= 0)
	{
		return;
	}
	Target.Hp = FMath::Max(0, Target.Hp - Damage);
	if (Target.Hp == 0)
	{
		Target.bDefeated = true;
	}
}

FAstralRoundOutcome UAstralBattleEngine::TryFinishRound()
{
	FAstralRoundOutcome Outcome;

	for (int32 Slot = 0; Slot < 2; Slot++)
	{
		FAstralCombatant* Member = GetActiveParty(Slot);
		if (Member && !Member->bDefeated && !bActed[Slot])
		{
			Outcome.bResolved = false;
			return Outcome;
		}
	}

	Outcome.Summary = ResolveOpponentActions();
	bActed[0] = bActed[1] = false;

	Outcome.bResolved = true;
	Outcome.bPlayerDefeat = Party ? AllDefeated(*Party) : false;
	return Outcome;
}

FString UAstralBattleEngine::ResolveOpponentActions()
{
	TArray<FString> Results;

	for (int32 i = 0; i < 2; i++)
	{
		FAstralCombatant& Enemy = Opponents[i];
		int32 TargetSlot = i % 2;
		FAstralCombatant* Target = GetActiveParty(TargetSlot);
		if (!Target || Target->bDefeated)
		{
			TargetSlot = (i + 1) % 2;
			Target = GetActiveParty(TargetSlot);
		}
		if (Enemy.bDefeated || !Target || Target->bDefeated)
		{
			continue;
		}

		const bool bWasGuarded = bGuarded[TargetSlot];
		const int32 Damage = UAstralCombatRules::ResolveIncomingDamage(OpponentDamage, bWasGuarded);
		Target->Hp = FMath::Max(0, Target->Hp - Damage);
		const bool bFainted = (Target->Hp == 0);
		if (bFainted)
		{
			Target->bDefeated = true;
		}

		Results.Add(FString::Printf(TEXT("%s%s took %d%s"),
			*Target->DisplayName,
			bWasGuarded ? TEXT(" guarded and") : TEXT(""),
			Damage,
			bFainted ? TEXT(" and fainted") : TEXT("")));
	}

	bGuarded[0] = bGuarded[1] = false;
	return Results.Num() == 0 ? TEXT("no counterattacks landed.") : FString::Join(Results, TEXT("; ")) + TEXT(".");
}

int32 UAstralBattleEngine::FindFirstEligibleParty(const TArray<FAstralCombatant>& Party, int32 Start)
{
	for (int32 i = FMath::Max(0, Start); i < Party.Num(); i++)
	{
		if (!Party[i].bDefeated)
		{
			return i;
		}
	}
	return -1;
}

bool UAstralBattleEngine::AllDefeated(const TArray<FAstralCombatant>& Party)
{
	for (int32 i = 0; i < Party.Num(); i++)
	{
		if (!Party[i].bDefeated)
		{
			return false;
		}
	}
	return true;
}
